examples/dataflow-hashpipeline/hashpipeline/buffering.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2020 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# pytype: skip-file
+
+from __future__ import absolute_import
+from apache_beam.transforms.userstate import BagStateSpec, TimerSpec, CombiningValueStateSpec, on_timer
+from apache_beam.transforms.timeutil import TimeDomain
+from apache_beam.coders import VarIntCoder, TupleCoder, StrUtf8Coder
+from apache_beam.transforms.combiners import CountCombineFn
+import apache_beam as beam
+import time
+from abc import ABC, abstractmethod
+from collections import defaultdict
+
+class BufferedPayloadFactory(ABC):
+  def __init__(self, max_size):
+    self.payloads = []
+    self.max_size = max_size
+
+  @abstractmethod
+  def running_payload(self):
+    '''Returns a singleton instance of the current payload'''
+    raise NotImplementedError
+  
+  def get_payloads(self):
+    '''Adds current payload to payloads list and flattens'''
+    self.payloads.append(self.running_payload())
+    return self.payloads
+  
+  @abstractmethod
+  def extend_running_payload(self, data):
+    '''Extends the current_singleton object until extending it further would pass max_size'''
+    raise NotImplementedError
+
+class FileLineBufferedPayloadFactory(BufferedPayloadFactory):
+  '''Payload factory for buffering lines in a file before sending batch requests to DLP'''
+  MAX_BYTE_SIZE_PER_REQUEST = 400000 # 0.4MB giving 100KB for metadata
+
+  def __init__(self, max_size = MAX_BYTE_SIZE_PER_REQUEST):
+    super(FileLineBufferedPayloadFactory, self).__init__(max_size)
+    self._current = ""
+  
+  def running_payload(self):
+    return self._current
+  
+  # NOTE: This does not handle the case where a single line is longer than max_size
+  def extend_running_payload(self, data):
```

Comment:
this implementation is confusing and buggy (see your note above and mine below).
It would be much more straightforward if the interface was familiar with the contract "add this data to the buffered payloads"

consider this refactor:
```python3
  @property
   def _available(self):
     self.max_size - len(self._current)

  def _close_current(self):
    self.payloads.append(self._current)
    self._current = ""

  def _avoid_ssn_split(data, isplit) -> int:
    #TODO(ryan canty): improve this
    while data[isplit] in {"-0123456789"} and isplit > 0:
      --isplit 
    return isplit

  def append(self, data):  # this replaces extend_running_payload
    if self._available == 0:
      self._close_current()
    if self._available > len(data):
      self._current += data
      return
    # prefer keeping as much of a line together as possible.
    if len(data) > self._available:
      self._close_current()
    isplit = self._available
    _avoid_split_ssn(data. isplit) 
    self._current += data[:isplit]
    self.append(data[isplit:]) # recurse to handle really long lines.
```

This now handles lines longer than max_size, doesn't drop lines and avoids splitting an SSN.